client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleInputGateDelegation.java

```diff
@@ -0,0 +1,565 @@
+
+  private List<InputChannelInfo> channelsInfo;
+  /** Map from channel index to reader index. */
+  private int[] channelIndexToReaderIndex;
+
+  /** Map from read index to channel index. */
+  private int[] readerIndexToChannelIndex;
+
+  /** The number of subpartitions that has not consumed per channel. */
+  private int[] numSubPartitionsNotConsumed;
+
+  /** The overall number of subpartitions that has not been consumed. */
+  private long numUnconsumedSubpartitions;
+
+  /** Received buffers from remote shuffle worker. It's consumed by upper computing task. */
+  private Queue<Pair<Buffer, InputChannelInfo>> receivedBuffers = new LinkedList<>();
+
+  /** {@link Throwable} when reading failure. */
+  private Throwable cause;
+
+  /** Whether this remote input gate has been closed or not. */
+  private boolean closed;
+
+  /** Whether we have opened all initial channels or not. */
+  private boolean initialChannelsOpened;
+
+  /** Number of pending {@link EndOfData} events to be received. */
+  private long pendingEndOfDataEvents;
+
+  /** Max concurrent reader count */
+  private int numConcurrentReading;
+
+  /** Keep compatibility with streaming mode. */
+  private boolean shouldDrainOnEndOfData = true;
+
+  /** Data decompressor. */
+  private BufferDecompressor bufferDecompressor;
+
+  private FlinkShuffleClientImpl shuffleClient;
+
+  private int numOpenedReaders = 0;
+  private AvailabilityProvider.AvailabilityHelper availabilityHelper;
+  private int startSubIndex;
+  private int endSubIndex;
+
+  public RemoteShuffleInputGateDelegation(
+      CelebornConf celebornConf,
+      String taskName,
+      int gateIndex,
+      InputGateDeploymentDescriptor gateDescriptor,
+      SupplierWithException<BufferPool, IOException> bufferPoolFactory,
+      BufferDecompressor bufferDecompressor,
+      int numConcurrentReading,
+      AvailabilityProvider.AvailabilityHelper availabilityHelper,
+      int startSubIndex,
+      int endSubIndex) {
+    this.taskName = taskName;
+    this.gateIndex = gateIndex;
+    this.gateDescriptor = gateDescriptor;
+    this.bufferPoolFactory = bufferPoolFactory;
+
+    int numChannels = gateDescriptor.getShuffleDescriptors().length;
+    this.channelIndexToReaderIndex = new int[numChannels];
+    this.readerIndexToChannelIndex = new int[numChannels];
+    this.numSubPartitionsNotConsumed = new int[numChannels];
+    this.bufferDecompressor = bufferDecompressor;
+
+    RemoteShuffleDescriptor remoteShuffleDescriptor =
+        (RemoteShuffleDescriptor) gateDescriptor.getShuffleDescriptors()[0];
+    this.shuffleClient =
+        FlinkShuffleClientImpl.get(
+            remoteShuffleDescriptor.getShuffleResource().getRssMetaServiceHost(),
+            remoteShuffleDescriptor.getShuffleResource().getRssMetaServicePort(),
+            celebornConf,
+            new UserIdentifier("default", "default"));
+
+    this.startSubIndex = startSubIndex;
+    this.endSubIndex = endSubIndex;
```

Comment:
flink 1.15 should support adaptive scheduler.
`
int startSubIdx = gateDescriptor.getConsumedSubpartitionIndexRange().getStartIndex();
int endSubIdx = gateDescriptor.getConsumedSubpartitionIndexRange().getEndIndex();
`